### PR TITLE
expression: fix the corner case of CAST int as unsigned real/decimal

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

In the original version, it will check inUnion && val<0 no matter whether the val is unsigned(the val is always int64, but in sql, it may be unsigned), so in some case, it will dirrerent from mysql5.7

This PR continue unfinished work in #11745.

### What is changed and how it works?

add check of whether unsigned

### Check List 

Tests 

 - Integration test

Code changes

 - None

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

Release note

 - None

### CODE REVIEW

**Comment by Deardrops** [comment]
@qw4990 PTAL

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -461,9 +461,10 @@ func (b *builtinCastIntAsRealSig) evalReal(row chunk.Row) (res float64, isNull b
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
+	unsigned := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
+	if !mysql.HasUnsignedFlag(b.tp.Flag) && !unsigned {
 		res = float64(val)
-	} else if b.inUnion && val < 0 {
+	} else if b.inUnion && !unsigned && val < 0 {
```
Comment (lonng) [inline expression/builtin_cast.go]: `if b.inUnion && unsigned && val < 0`?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -487,9 +488,10 @@ func (b *builtinCastIntAsDecimalSig) evalDecimal(row chunk.Row) (res *types.MyDe
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
+	unsigned := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
```
Comment (lonng) [inline expression/builtin_cast.go]: ditto

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -487,9 +488,10 @@ func (b *builtinCastIntAsDecimalSig) evalDecimal(row chunk.Row) (res *types.MyDe
...
+	if !mysql.HasUnsignedFlag(b.tp.Flag) && !unsigned {
 		res = types.NewDecFromInt(val)
-	} else if b.inUnion && val < 0 {
+	} else if b.inUnion && !unsigned && val < 0 {
```
Comment (lonng) [inline expression/builtin_cast.go]: ditto

File: expression/builtin_cast.go (hunk 2 above)
Comment (lonng) [inline expression/builtin_cast.go]: ditto

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -461,9 +461,10 @@ func (b *builtinCastIntAsRealSig) evalReal(row chunk.Row) (res float64, isNull b
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
+	unsigned := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
```
Comment (lonng) [inline expression/builtin_cast.go]: It's better to define the `unsigned` in the `IF` statement to restrict its scope. e.g `if unsigned := ...; unsigned... {}`

Comment (lonng) [inline expression/builtin_cast.go]: It's better to define the `unsigned` in the `IF` statement to restrict its scope. e.g `if unsigned := ...; unsigned... {}`

File: expression/builtin_cast.go (hunk 1 above)
Comment (lonng) [inline expression/builtin_cast.go]: I think `} else if b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && val < 0 {` is more comprehensable.

File: expression/builtin_cast_vec.go [hunk 5]
```diff
@@ -109,7 +109,7 @@ func (b *builtinCastIntAsRealSig) vecEvalReal(input *chunk.Chunk, result *chunk.
 		}
 		if !hasUnsignedFlag0 && !hasUnsignedFlag1 {
 			rs[i] = float64(i64s[i])
-		} else if b.inUnion && i64s[i] < 0 {
+		} else if b.inUnion && !hasUnsignedFlag1 && i64s[i] < 0 {
```
Comment (lonng) [inline expression/builtin_cast_vec.go]: ditto

File: expression/builtin_cast.go (hunk 4 above)
Comment (SunRunAway) [inline expression/builtin_cast.go]:
```suggestion
	unsignedArgs0 := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
```

File: expression/builtin_cast_vec.go (hunk 5 above)
Comment (SunRunAway) [reply expression/builtin_cast_vec.go]: It's in a fop loop, maybe it's acceptable here.

**Comment by Deardrops** [comment]
@SunRunAway Comment addressed, PTAL

File: expression/builtin_cast_vec.go (hunk 5 above)
Comment (lonng) [reply expression/builtin_cast_vec.go]: @SunRunAway I am not mean the local variable, see: https://github.com/pingcap/tidb/pull/13637#discussion_r349977712, I mean this logic is not comprehensable.

Comment (SunRunAway) [reply expression/builtin_cast_vec.go]: I do not get you. What is your suggestion?

File: expression/builtin_cast.go [hunk 6]
```diff
@@ -461,9 +461,10 @@ func (b *builtinCastIntAsRealSig) evalReal(row chunk.Row) (res float64, isNull b
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
+	if unsignedArgs0 := mysql.HasUnsignedFlag(b.args[0].GetType().Flag); !mysql.HasUnsignedFlag(b.tp.Flag) && !unsignedArgs0 {
```
Comment (lonng) [inline expression/builtin_cast.go]: How about:
```go
	} else if b.inUnion && !unsignedArgs0 && val < 0 {
		// Round up to 0 if the value is negative but the expression eval type is unsigned in `UNION` statement
		// NOTE: the following expressions are equal (so choose the more efficient one):
		// `b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && !unsignedArgs0 && val < 0`
		// `b.inUnion && !unsignedArgs0 && val < 0`
		res = 0
	} else {
```

**Pull request merged**